### ISSUE TITLE
Dedupe of uploaded records

### ISSUE THREAD

**Issue by lmullane** [bcgov/foi-flow#3120]
* As a ministry coordinator
* I want to be informed on what documents are deduped and converted to PDF when I upload a batch of records
* so that I know what has happened with the upload

**Assumptions & Scope**	
What are the assumptions for this story?

Removing the record from the user interface does not mean removing it from the Object Storage; the original will always be kept in Object Storage. 
Ministry coordinator will upload the documents in batches as they are received from the divisions.

IAO will not upload but will still have the ability to do so.

Only works with PDFs currently; other record types will be covered in #3205

#3205 must be ready for QA before testing Scenario 5.
#3136 must be ready for QA before testing Scenario 7. 

What is IN scope?

What is NOT in scope?
Notification that an upload is complete. #3119
UI for file upload status is complete, but will not be updated with correct values since that depends on #3119

Search and filter by division in scope. 

**Acceptance Criteria**	

**Scenario 1: Dedupe and in progress**
* GIVEN a ministry coordinator has added a batch of records to the upload records modal
* WHEN the coordinator clicks 'Continue'
* THEN the modal will close
* AND an uploading icon will show next to each new file
* AND the a state will be with each file saying, "Deduplication and file conversion in progress."
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/957399281

**Scenario 2: Dedupe is complete**

- GIVEN a batch of records have been uploaded
- WHEN the deduplication is complete
- THEN the uploaded records will be listed
- AND green check mark will be next to each unique record
- AND a yellow duplicate symbol next to each removed duplicate

**Scenario 3: Duplicate of this document message**

- GIVEN a record has been uploaded
- WHEN a record has been removed as a duplicate
- THEN the statement next to the record will say, "Duplicate of (NAME)"

**Scenario 4: Duplicate of this document message**
* GIVEN a record has been uploaded
* WHEN a record is unique
* THEN the statement next to the record will say, "Ready for redaction"
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/975255939

**Scenario 5: Number of batches uploaded**

- GIVEN a batch of record has been uploaded
- WHEN dedupe and file conversion is in progress or complete on the batch
- THEN at the top of the page the ''Batches Uploaded' will be increase: 1, 2, 3, etc.
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/975255939

**Scenario 6: Number of files deduplicated**

- GIVEN a batch of record has been uploaded
- WHEN a record(s) has been deduplicated
- THEN at the top of the page the 'Files Uploaded' will increase: 23, 32, 37 etc. https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/975255939

**Scenario 7: Delete Deduplicated Files - Update Counter**

- GIVEN a ministry coordinator or IAO analyst uploads a duplicate record
- AND the record has been deduplicated (removed as duplicate)
- WHEN the ministry coordinator or IAO analyst deletes the deduplicated record (the duplicate)
- THEN the 'Deduplicated Files" counter will decrease   

**Dependencies? What is the impact of this dependency? (If so, link dependency in the ticket, make it visible in a teams backlog)**

**Validation Rules? (If yes, list here)**	

**Design**
@xxx - please link the Design here

**Definition of Ready**

1. [ ] Is there a well articulated User Story?
2. [ ] Is there Acceptance Criteria that covers all scenarios (happy/sad paths)?
3. [ ] If there is a user interface, is there a design? 
4. [ ] Does the user story need user research/validation?
5. [ ] Does this User Story needs stakeholder approval?
6. [ ] Design / Solution accepted by Product Owner
7. [ ] Is this user story small enough to be completed in a Sprint? Should it be split?
8. [ ] Are the dependencies known/ understood? (technical, business, regulatory/policy)
9. [ ] Has the story been estimated?

**Definition of Done**

1.    [ ] Passes developer unit tests
2.    [ ] Passes peer code review
3.    [ ] If there's a user interface, passes UX assurance
4.    [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
5.	[ ] Confirm Test cases built and succeeding 
6.	[ ] No regression test failures
7.	[ ] Test coverage acceptable by Product Owner
8.	[ ] Ticket ready to be merged to master or story branch
9.	[ ] Developer to list Config changes/ Update documents and designs
10.	[ ] Can be demoed in Sprint Review 
11.	[ ] Tagged as part of a Release
12.   [ ] Feature flagged if required
13.  [ ] Change Management activities done?

**Pre-Deployment Steps**
 do the following in doc reviewer app
- [ ] set env variable in open shift "OSS_S3_ENVIRONMENT" to dev/test/prod
- [ ] add buckets in s3 for each ministry according to the names DocumentPathMapper
- [ ] run the following query for all buckets 
```
update public."DocumentPathMapper" set attributes = '{
         "s3accesskey":"** enter access key **",
         "s3secretkey":"** enter secret key **"
     }' where bucket ilike '** enter bucket name here **%'
```
- [ ] add the following env variables in docreviewer app
```
DEDUPE_REDIS_HOST=
DEDUPE_REDIS_PASSWORD=
DEDUPE_REDIS_PORT=
DEDUPE_STREAM_KEY=FOIDEDUPE
DEDUPE_S3_ENV=dev ## dev | test | prod  - this is as suffix on bucket name

DEDUPE_DB_HOST=
DEDUPE_DB_NAME=
DEDUPE_DB_PORT=
DEDUPE_DB_USER=
DEDUPE_DB_PASSWORD=

DEDUPE_S3_HOST=
DEDUPE_S3_REGION=
DEDUPE_S3_SERVICE=
```
- [ ] make sure docreviewer api changes are deployed first, as foi flow api changes depends on endpoint in reviewer api
- [ ] add the following env variables to request-managemnet-api in dev/test/prod

```
#event queue
EVENT_QUEUE_HOST=enter host here
EVENT_QUEUE_PORT=enter port here
EVENT_QUEUE_PASSWORD=password
EVENT_QUEUE_CONVERSION_STREAMKEY='file-conversion'
EVENT_QUEUE_DEDUPE_STREAMKEY='FOIDEDUPE'

#Doc Reviewer URL
FOI_DOCREVIEWER_BASE_API_URL=enter api here
```

**Comment by lmullane** [comment]
@abin-aot, @JHarrietha-AOT 

Please review. 

@m-prodan, does IAO staff usually say "files" or "records"? Want to ensure we have the language right, when we say "Number of Files Removed"

cc @KyEggleston @liseandtea

**Comment by m-prodan** [comment]
It is currently tracked as number of pages deduplicated, which as I understand is largely an estimate. I think it is based off the number of files removed as duplicate by Intella. 

Is it possible to track the number of pages deduplicated @abin-aot?

**Comment by JHarrietha-AOT** [comment]
@lmullane AC Looks good to me

@m-prodan Is this to ensure we accurately represent page count? I would assume our page count would be calculated based off of the number of pages that actually end up in the redaction app after this has been completed?
Or is this because Analyst often think in page count so it's a better representation of the time they would save? (Also how do we have it worded in out KPI's? Page count or documents?)

cc: @abin-aot just my thoughts on Matt's question above

**Comment by m-prodan** [comment]
Good point - in the KPI it's actually records, but in AXIS it is tracked as pages deduplicated. A record could be 1 to 100+ pages so how we are currently tracking in AXIS doesn't necessarily translate well. Analysts do often think in page counts so they would perceive the time savings better if they saw the number of pages. Ideally it could say xx number of records deduplicated containing xx pages

**Comment by JHarrietha-AOT** [comment]
If we create a report to show this in the future would we also want to show both? Like I think its valid to say the number of documents don't mean alot to users and they could see that through the visual icons in the Gather Records longs, so then at the top we just show pages?

@liseandtea curious what numbers would mean more to you as a person who was most recently an analyst

**Comment by lmullane** [comment]
@m-prodan, I'm sure Abin could capture the number of pages deduplicated. Do we need to show that on the user interface or only capture that in reporting?

**Comment by JHarrietha-AOT** [comment]
Adding a screenshot of some minor design updates I have run past @aparna-aot who is working on the front end of this ticket. We are now showing the size of the file based off of users requests. And we are shifting the "status" of the document and the username/date & time. ![image.png](https://images.zenhubusercontent.com/6081a3ba692e484184671a59/eb5178e2-9c1a-435f-815d-7afede77a48c)

**Comment by KyEggleston** [comment]
@nkan-aot - I ran through the ACs, and here are the results - let me know if you need any more information. 

<h3>Scenario 2 - Tested on EDU-2022-12396 and ABC-3453453-435345</h3>

Current functionality appears to have dedupe looking at the files contained in ALL requests, rather than just the request you are uploading files to, and if a duplicate is found in ANY request, it will mark the file as a duplicate on your current request.

E.g., I uploaded a unique file called "Vulnerability Assessment.pdf" to a request, but since that exact same file was already uploaded to another request, it marked it as a duplicate. 

I think we've brought this up already, so it may already be known/being looked into. 

<h3>Scenario 3 - Tested on EDU-2022-12396 and ABC-3453453-435345</h3>

Technically this passed, but the issue in Scenario 2 prevents accurate dedupe results from occurring (as non-duplicates on the request are being removed as duplicates). 

<h3>Scenario 5 - Tested on EDU-2022-12396 and ABC-3453453-435345</h3>

Currently, dedupe and file conversion does not need to be completed (in full) in order for the 'Batches Uploaded' to be updated (i.e., to go from 0 to 1).

The 'Batches Uploaded' counter is updated even if a file for the current batch is in the process of dedupe and file conversation. Not a huge deal, and perhaps this is fine and the AC can be updated. 

<h3>Scenario 7 - Tested on ABC-3453453-435345</h3>

Unable to delete any file type, in any status. Clicked the elipsis -> delete, but the continue button is grey and inactive. In addition, unable to confirm the 'Number of deleted files' is updated, as I'm unable to delete files, and there is no counter with this title (unless we are referring to 'Files Removed', in which case I can update the AC).

<h3>Other issues</h3>

I've noticed intermittent issues when uploading files to a request. The 'Successfully Saved' toast comes up after clicking 'Continue' on the Upload Records modal (when records are added), but the records are not uploaded to the Records tab. 

In addition, I have found situations where I successfully upload files to a request, and they were added to the _Records_ tab, but after returning to the request after an undetermined amount of time, they had disappeared and were no longer displayed. 

These issues above both occurred on EDU-2022-12396, which was one of the requests I tested the ACs on. So I retested the ACs with ABC-3453453-435345 to rule out an issue with that specific request.

cc: @lmullane @m-prodan @liseandtea

**Comment by nkan-aot** [comment]
hey @KyEggleston 
For scenario 2, that is fixed and pushed to dev-marshal but currently we are still in the process of setting up the test reviewer in the proper openshift project, so it is not available in test-marshal yet. Feel free to test it in dev-marshal and verify it is behaving correctly

For scenario 5, we do not currently have a way to tell when a single batch is finished deduplicating, only individual files. This is being addressed in ticket #3205, I will update the behaviour once that is complete

For scenario 7, delete has not been implemented yet, this will be addressed in #3136. Also was under the impression that a user deleted file and a file removed from deduplication are two different concepts. A user deleted file means it is no longer responsive and no longer needs to be displayed, while a deduplicated file is still responsive, and we are  showing the number up top to specifically indicate how many files were removed during dedupe process. If we mix these two numbers then it becomes hard to tell what was removed by a user vs dedupe. We can clarify this tmr, and if user deleted files need to be included in the count then it will be fixed after #3136

I can confirm the record disappearing problem is happening in test-marshal but unfortunately I cannot identify the cause in test-marshal and I cannot replicate the issue in dev-marshal or locally. It may be the case that new code pushed to dev-marshal has already fixed the issue, I will keep an eye on it next time test-marshal is deployed.

**Comment by KyEggleston** [comment]
Thanks @nkan-aot 

For Scenario 2, I don't have access to dev-marshal, so I cant test it at the moment. But that's ok, and I can wait for it to be in test-marshal.

For Scenario 5 and 7, I will wait for the respective tickets (#3205 and #3136) for each feature to be ready for QA. At that time I may update these ACs if needed. I added these ticket numbers to the original post, indicating that they must be ready for QA to test these scenarios. 

You're right that a user deleted file and a file removed from deduplication are two different concepts (confirmed at our stand up today). User deleted files do not need to be displayed (i.e., added to the 'Files Removed' counter). Files removed through deduplication would be added to the 'Files Removed' counter (which is currently occurring).  

For the other issue(s) I identified (regarding the records disappearing problem), let me know if the issue does come up in test-marshal next deployment, and I can make a bug ticket if needed. 

Should this ticket be moved back to Review until we can test all the ACs?

**Comment by lmullane** [comment]
@nkan-aot, as discussed, 'Files Removed' should be changed to 'File Deduplicated.' This counter should only include files that are removed as duplicated, not deleted.

The first counter for 'Deduplicated Files' should be changed to 'Files Uploaded.'

**Comment by KyEggleston** [comment]
All ACs pass, and updated ACs to align with recent changes/discussions.
